### ISSUE TITLE
Custom JsonConverter not used for deserialization of items Reported by IProgress instance.

### ISSUE THREAD

**Issue by NickRedwood** [microsoft/vs-streamjsonrpc#377]
The context of this is that I have been trying to use an instance of `IProgress<System.Reactive.Notification<T>>` as a roughly isomorphic proxy for `IObservable<T>`. `System.Reactive.Notification<T>` requires custom serialization/deserialization because it's immutable and has properties that throw errors in some cases. After writing this serialization code (a `JsonConverter`) and passing this in to server and client, I found that deserialization still won't work correctly when passing an instance `IProgress<Reactive.Notification<T>>` into the RPC call. 

The issue seems to be that the custom `JsonConverter` is not called on the receiving side to deserialize the message. Perhaps a default `JsonSerializer` instance is used within reported value deserialization instead? The simpler example without `Notification<T>` code below demonstrates this:

```cs
public class ServerWithProgress
{
	public Task<int> GetProgress(IProgress<int> p)
	{
		Console.WriteLine($"Received request with Progress instance");
		p.Report(1);
		return Task.FromResult(10);
	}

}    
```

```cs
public class PlusOneJsonConverter : JsonConverter<int>
{
	public override int ReadJson(JsonReader reader, Type objectType, int existingValue, bool hasExistingValue, JsonSerializer serializer)
	{
		var val = JsonSerializer.CreateDefault().Deserialize<int>(reader);
                // Just to show that this method has or has not been called.
		return val + 1;
		
	}

	public override void WriteJson(JsonWriter writer, int value, JsonSerializer serializer)
	{
		JsonSerializer.CreateDefault().Serialize(writer,value);
	}
}
```

```cs
async Task Main()
{
	var converter = new PlusOneJsonConverter();
	var (clientStream, serverStream) = Nerdbank.Streams.FullDuplexStream.CreatePair();

	var serverFormatter = new JsonMessageFormatter();
	serverFormatter.JsonSerializer.Converters.Add(converter);
	var serverRpc = new JsonRpc(new HeaderDelimitedMessageHandler(serverStream, serverFormatter),new ServerWithProgress());

	var clientFormatter = new JsonMessageFormatter();
	clientFormatter.JsonSerializer.Converters.Add(converter);
	var clientRpc = new JsonRpc(new HeaderDelimitedMessageHandler(clientStream, clientFormatter));

	serverRpc.StartListening();
	clientRpc.StartListening();

	var val = 0;
	var p = new ProgressWithCompletion<int>(i => { Console.WriteLine("Progress received " + i); val = i;});
	
	var t =  clientRpc.InvokeAsync<int>("GetProgress", p);
	
	await p.WaitAsync();
	var final = await t;
	Console.WriteLine("Val is " + val);
	Console.WriteLine("Final value is " + final);
        // Fails as val is still 1
	Debug.Assert(val == 2);
}
```

The final value returned _does_ use the custom converter `ReadJson` when deserialized, while the progress value _does not_.

I can create a surrogate version of System.Reactive.Notification<T> that won't require a custom `JsonConverter`, plus would have to do the same for all T types. Is there anything else that can be done or, if this is a bug, can it be fixed?

**Comment by AArnott** [comment]
Thanks for the great report. We'll get this fixed in a v2.2 servicing release.

**Comment by NickRedwood** [comment]
Great, thanks Andrew
